src/Control/Monad/Logic/Sequence.hs

```diff
@@ -165,17 +173,17 @@ observeT :: Monad m => SeqT m a -> m a
 observeT :: MonadFail m => SeqT m a -> m a
 #endif
 observeT (toView -> m) = m >>= go where
-  go (Just (a, _)) = return a
-  go _ = fail "No results."
+  go (VCons a _) = return a
+  go VEmpty = fail "No results."
 
 observe :: Seq a -> a
 observe (toView -> m) = case runIdentity m of
-  Just (a, _) -> a
-  _ -> error "No results."
+  VCons a _ -> a
+  VEmpty -> error "No results."
 
 observeMaybeT :: Monad m => SeqT m (Maybe a) -> m (Maybe a)
 observeMaybeT (toView -> m) = m >>= go where
-  go (Just (Just a, _)) = return (Just a)
```

Comment:
I translated `observeMaybeT` from some other code and I think I wrote it wrong. The nothing case is supposed to try the rest. It's basically supposed to filter off any leading `Nothing`.

Reply:
I'll have a look.

Reply:
Before your changes, I think this is roughly what I meant to write:
```haskell
observeMaybeT :: Monad m => SeqT m (Maybe a) -> m (Maybe a)
observeMaybeT (toView -> m) = m >>= go where
  go (Just (Just a, _)) = return (Just a)
  go (Just (Nothing, rest)) = toView rest >>= go
  go _ = return Nothing
```

Reply:
I don't think the name is great. I think of this as `catMaybes` up to the first `Just`. And it definitely needs some clear documentation as to why/how this is different than just `observeT` when the computation returns `Maybe`. It was useful to me recently for something where I was already returning `Maybe`s but I'm not sure it's all that generally useful because I think you could always just do `catMaybes . observeFoo` possible with a `take N` in there, for a suitable observe function.